File: recipes/subms-stats/example/java/src/main/java/com/submillisecond/primers/stats/StatsMain.java
package com.submillisecond.primers.stats;

import com.submillisecond.stats.Bootstrap;

import java.util.Optional;

/**
 * Full primer report. Generates a baseline + a heavy-tailed candidate,
 * prints {@link StatsReport} blocks for each, and runs an A/B compare
 * (KS + Cohen's d) plus a 95% bootstrap CI on the candidate's p99.
 *
 * <p>Run with: {@code java -cp target/classes com.submillisecond.primers.stats.StatsMain}.
 */
public final class StatsMain {

    private static final int SAMPLE_COUNT = 50_000;
    private static final long BASE_NS     = 800_000L;     // ~800us nominal latency
    private static final long BASELINE_SEED  = 1L;
    private static final long CANDIDATE_SEED = 2L;

    private StatsMain() {}

    public static void main(String[] args) {
        long[] baseline = Workload.generate(
                SAMPLE_COUNT, BASE_NS, Workload.TailShape.CLEAN, BASELINE_SEED);
        long[] candidate = Workload.generate(
                SAMPLE_COUNT, BASE_NS, Workload.TailShape.HEAVY, CANDIDATE_SEED);

        StatsReport baseReport = Analyser.analyse("baseline", baseline);
        StatsReport candReport = Analyser.analyse("candidate", candidate);

        printReport(baseReport);
        printReport(candReport);
        printCompare(baseline, candidate);
        printBootstrap(candReport);
    }

    private static void printReport(StatsReport r) {
        System.out.println();
        System.out.println("== " + r.label() + " (n=" + r.count() + ") ==");
        System.out.printf("  percentiles    p50=%dns  p90=%dns  p99=%dns  p99.9=%dns  max=%dns%n",
                r.p50(), r.p90(), r.p99(), r.p999(), r.max());
        System.out.printf("  moments        mean=%dns  stddev=%dns  cv=%.3f  skew=%.3f  kurt=%.3f%n",
                r.mean(), r.stddev(), r.cv(), r.skewness(), r.kurtosis());
        System.out.printf("  robust spread  iqr=%dns  mad=%dns%n",
                r.iqr(), r.mad());
        System.out.printf("  tail           cte99=%dns  fatness(p99/p50)=%.2f  hill(k=50)=%s%n",
                r.cte99(), r.tailFatness(), formatHill(r.hillIndex()));
        System.out.printf("  rig stability  jitterScore=%.3f%n",
                r.jitterScore());
        System.out.printf("  bootstrap      p99 95%% CI = [%dns, %dns]%n",
                r.p99Ci().lo(), r.p99Ci().hi());
    }

    private static void printCompare(long[] baseline, long[] candidate) {
        Optional<Double> ks = Analyser.ks(baseline, candidate);
        Optional<Double> d  = Analyser.cohensD(baseline, candidate);
        System.out.println();
        System.out.println("== compare: candidate vs baseline ==");
        System.out.printf("  KS statistic  = %s   (0.0 identical CDFs, 1.0 disjoint)%n",
                ks.map(v -> String.format("%.4f", v)).orElse("n/a"));
        System.out.printf("  Cohen's d     = %s   (sign: positive = candidate slower)%n",
                d.map(v -> String.format("%+.4f", v)).orElse("n/a"));
    }

    private static void printBootstrap(StatsReport candidate) {
        Bootstrap.CI ci = candidate.p99Ci();
        System.out.println();
        System.out.println("== bootstrap CI (candidate p99, 95%, 500 iters, seed=0) ==");
        System.out.printf("  p99   = %dns%n", candidate.p99());
        System.out.printf("  95%% CI = [%dns, %dns]   width = %dns%n",
                ci.lo(), ci.hi(), Math.max(0L, ci.hi() - ci.lo()));
    }

    private static String formatHill(Optional<Double> hill) {
        return hill.map(v -> String.format("%.3f", v)).orElse("n/a");
    }
}

File: recipes/subms-stats/example/rust/examples/stats_main.rs
//! End-to-end driver for the `subms-stats` recipe example. Builds two
//! synthetic latency-sample streams - a "baseline" with bounded outliers and
//! a "candidate" with a deliberate power-law tail regression - and prints a
//! structured analysis for each, followed by a two-run comparison block.
//!
//! ```sh
//! cargo run --release --example stats_main
//! ```
//!
//! Output is plain text. No machine-readable JSON is emitted; that's the
//! job of a recipe's `perf_main`, not this example.

use subms_stats_example::{StatsReport, TailShape, Workload, analyse, compare};

fn main() {
    let baseline_wl = Workload {
        n: 5_000,
        base_ns: 800,
        shape: TailShape::Uniformish,
        seed: 0xBA5E_BA5E_BA5E_BA5E,
    };
    let candidate_wl = Workload {
        n: 5_000,
        base_ns: 850,
        shape: TailShape::PowerLaw,
        seed: 0xCADD_A7E0_CADD_A7E0,
    };

    let baseline = baseline_wl.generate();
    let candidate = candidate_wl.generate();

    print_workload("baseline", &baseline_wl);
    let baseline_report = analyse(&baseline);
    print_report(&baseline_report);

    print_workload("candidate (slight regression)", &candidate_wl);
    let candidate_report = analyse(&candidate);
    print_report(&candidate_report);

    println!();
    println!("=== Compare (baseline vs candidate) ===");
    let cmp = compare(&baseline, &candidate);
    let ks = cmp.ks.map(|v| format!("{v:.4}")).unwrap_or_else(empty);
    let d = cmp
        .cohens_d
        .map(|v| format!("{v:+.4}"))
        .unwrap_or_else(empty);
    println!("  KS statistic           : {ks}");
    println!("  Cohen's d              : {d}");
    println!(
        "  baseline  p99 95% CI   : [{}, {}] ns",
        cmp.baseline_p99_ci.0, cmp.baseline_p99_ci.1
    );
    println!(
        "  candidate p99 95% CI   : [{}, {}] ns",
        cmp.candidate_p99_ci.0, cmp.candidate_p99_ci.1
    );

    // Editorial verdict. Threshold values track the conventional Cohen's d
    // ranges (small / medium / large) and a KS gap that won't fire on
    // run-to-run noise at n=5k.
    let verdict = match (cmp.ks, cmp.cohens_d) {
        (Some(ks), Some(d)) if ks > 0.10 && d.abs() > 0.20 => {
            "distribution shift detected (KS clears 0.10, |d| clears 0.20)"
        }
        (Some(_), Some(_)) => "no clear shift - changes within noise",
        _ => "comparison unavailable (empty input?)",
    };
    println!("  verdict                : {verdict}");
}

fn print_workload(label: &str, w: &Workload) {
    println!();
    println!("=== Run: {label} ===");
    println!(
        "  workload               : n={} base_ns={} shape={:?} seed={:#x}",
        w.n, w.base_ns, w.shape, w.seed
    );
}

fn print_report(r: &StatsReport) {
    println!("  count                  : {}", r.count);
    println!(
        "  p50 / p90 / p99 / p99.9 : {} / {} / {} / {} ns",
        r.p50_ns, r.p90_ns, r.p99_ns, r.p999_ns
    );
    println!("  max                    : {} ns", r.max_ns);
    println!(
        "  mean / stddev          : {} / {} ns",
        r.mean_ns, r.stddev_ns
    );
    println!(
        "  p99 95% CI (bootstrap) : [{}, {}] ns",
        r.p99_ci_ns.0, r.p99_ci_ns.1
    );
    println!(
        "  CTE@p99                : {} ns   (mean of samples above p99)",
        r.cte99_ns
    );
    println!("  tail fatness (p99/p50) : {:.3}", r.tail_fatness);
    match r.hill_50 {
        Some(h) => println!("  Hill index (k=50)      : {h:+.3}"),
        None => println!("  Hill index (k=50)      : <too few samples>"),
    }
    println!("  IQR / MAD              : {} / {} ns", r.iqr_ns, r.mad_ns);
    println!(
        "  CoV / skew / kurt      : {:.3} / {:+.3} / {:+.3}",
        r.cov, r.skew, r.kurt
    );
    println!("  jitter (0=clean,1=loud): {:.3}", r.jitter);

    // Percentile sweep - one quick scan from p50 to p999 in 10-percentile
    // steps, so the body and the climb to the tail land on the same line.
    let sweep: String = r
        .sweep
        .iter()
        .map(|(q, ns)| format!("p{:>5.1}={}ns", q * 100.0, ns))
        .collect::<Vec<_>>()
        .join("  ");
    println!("  sweep                  : {sweep}");

    // CDF buckets: log2-spaced. Print only the non-zero stretch so the
    // output stays readable.
    let first = r.cdf_buckets.iter().position(|&c| c > 0).unwrap_or(0);
    let last = r.cdf_buckets.iter().rposition(|&c| c > 0).unwrap_or(first);
    let mut hist = String::new();
    for (i, &c) in r.cdf_buckets[first..=last].iter().enumerate() {
        let bucket = first + i;
        let lo = if bucket == 0 { 0 } else { 1u64 << bucket };
        let hi = 1u64 << (bucket + 1);
        hist.push_str(&format!("  [{lo}..{hi}):{c}"));
    }
    println!("  CDF buckets            :{hist}");
}

fn empty() -> String {
    "<n/a>".to_string()
}
